### PR TITLE
User Menu component

### LINKED ISSUE #4: Component: User Menu
Main UI is an avatars (using Gravatar, with https://ui-avatars.com/ or somehing similar as a fallback)

When clicking it, opens something dropdown-ish containing:

- Login/Logout (just show one for now and keep the other one hidden - we'll update that once the login code is ready)
- Your full name/email?

---

Or do we want to show something completely different while not logged-in? Probably something to see later; for now having a Login action in there is the easiest option.

### PR DESCRIPTION
**Pull request by mic4ael**
closes #4

### CODE REVIEW

File: newdle/client/package.json [hunk 1]
```diff
@@ -3,11 +3,17 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "chai": "^4.2.0",
+    "enzyme": "^3.10.0",
```
Comment (ThiefMaster) [inline newdle/client/package.json]: that's pretty certainly a dev dependency

File: newdle/client/src/UserMenu.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  '8b5d5d', // color
+  2, // length
+  0.5, // font size
+  true, // rounded
+];
+
+export default function UserMenu({user}) {
+  const {email, fullName} = user;
+  const gravatarFallback = fullName
+    ? encodeURI(`//ui-avatars.com/api/${fullName}/${defaultUIAvatarsArgs.join('/')}`)
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.jsx]: I think it's enough to pass the first character of the first/last name here; No need to expose more data than needed to an external service.

File: newdle/client/package.json (hunk 1 above)
Comment (mic4ael) [reply newdle/client/package.json]: Yeah, indeed. I was to do it when cleaning up the code ;)

File: newdle/client/src/UserMenu.jsx (hunk 2 above)
Comment (mic4ael) [reply newdle/client/src/UserMenu.jsx]: Yep, definitely ;)

File: newdle/client/package.json [hunk 3]
```diff
@@ -5,13 +5,19 @@
...
-    "react-scripts": "3.0.1"
+    "react-gravatar": "^2.6.3",
+    "react-scripts": "3.0.1",
+    "semantic-ui-react": "^0.87.3"
   },
   "devDependencies": {
     "flask-urls.macro": "0.0.1",
-    "node-sass": "^4.12.0"
+    "node-sass": "^4.12.0",
+    "chai": "^4.2.0",
```
Comment (ThiefMaster) [inline newdle/client/package.json]: why do we need to explicitly depend on it? i can't see it referenced anywhere in our code. is it a peer dependency?

File: newdle/client/src/UserMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  );
+}
+
+UserMenu.propTypes = {
+  user: PropTypes.shape({
+    email: PropTypes.string,
+    fullName: PropTypes.string,
+    firstName: PropTypes.string,
+    lastName: PropTypes.string,
+  }).isRequired,
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.jsx]: Won't this fail for `user={null}`?

File: newdle/client/src/UserMenu.test.jsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  it('renders UserMenu successfully for anonymous users', () => {
+    const anonymous = mount(<UserMenu />);
+    expect(anonymous.contains(Icon)).to.equal(true);
+    expect(anonymous.exists('.user-gravatar')).to.equal(false);
+    anonymous.unmount();
+  });
+
+  it('renders UserMenu successfully for users with Gravatar', () => {
+    const userData = {email: 'test@test.email', fullName: 'T T', firstName: 'T', lastName: 'T'};
+    const anonymous = mount(<UserMenu user={userData} />);
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.test.jsx]: this isn't *anonymous* ;)

File: newdle/client/package.json (hunk 3 above)
Comment (mic4ael) [reply newdle/client/package.json]: There is one reference. `UserMenu.test.js`

File: newdle/client/src/UserMenu.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.js]: why the empty line here?

File: newdle/client/src/UserMenu.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        <span>{fullName || email}</span>{' '}
+        <Gravatar email={email} default={gravatarFallback} size={40} />
+      </div>
+    )
+  );
+}
+
+export default function UserMenu() {
+  const user = useSelector(getUserInfo);
+  const trigger = user ? renderGravatar(user) : <NoUserAvatar />;
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.js]: I'd move this after all the hooks

File: newdle/client/src/UserMenu.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+function renderGravatar({email, first_name: firstName, last_name: lastName, full_name: fullName}) {
+  const gravatarFallback =
+    firstName && lastName
+      ? encodeURI(
+          `//ui-avatars.com/api/${firstName[0]} ${lastName[0]}/${defaultUIAvatarsArgs.join('/')}`
+        )
+      : 'mp';
+
+  return (
+    email && (
```
Comment (ThiefMaster) [inline newdle/client/src/UserMenu.js]: `if (!email) return null;` at the beginning.. the fallback defined above isn't even used then.

However, I don't think we actually need it: If we have a user we always have an email as well.

**Pull request merged**